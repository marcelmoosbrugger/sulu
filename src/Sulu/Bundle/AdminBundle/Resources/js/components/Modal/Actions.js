// @flow
import React from 'react';
import type {Action} from './types';
import actionsStyles from './actions.scss';

type Props = {
    actions: Array<Action>,
};

export default class Actions extends React.PureComponent<Props> {
    render() {
        if (this.props.actions.length > 0) {
            return (
                <div className={actionsStyles.actions}>
                    {this.props.actions.map((action, index) => {
                        const handleButtonClick = action.onClick;
                        return (
                            <button
                                key={index}
                                className={actionsStyles.action}
                                onClick={handleButtonClick}>{action.title}</button>
                        );
                    })}
                </div>
            );
        }

        return null;
    }
}
