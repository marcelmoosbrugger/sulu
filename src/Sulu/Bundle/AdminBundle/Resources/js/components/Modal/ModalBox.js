// @flow
import React from 'react';
import Icon from '../Icon';
import type {ModalProps, Action} from './types';
import modalBoxStyles from './modalBox.scss';

type Props = ModalProps & {
    onRequestClose: () => void,
}

const CLOSE_ICON = 'times';

export default class ModalBox extends React.PureComponent<Props> {
    static defaultProps = {
        actions: [],
    };

    render() {
        return (
            <div className={modalBoxStyles.box}>
                <div className={modalBoxStyles.header}>
                    {this.props.title}
                    <Icon
                        name={CLOSE_ICON}
                        className={modalBoxStyles.icon}
                        onClick={this.props.onRequestClose} />
                </div>
                <div className={modalBoxStyles.content}>
                    {this.props.children}
                </div>
                <div className={modalBoxStyles.footer}>
                    {ModalBox.renderActions(this.props.actions)}
                    <button
                        className={modalBoxStyles.confirmButton}
                        onClick={this.props.onConfirm}>{this.props.confirmText}</button>
                </div>
            </div>
        );
    }

    static renderActions(actions: Array<Action>) {
        if (actions.length > 0) {
            return (
                <div className={modalBoxStyles.actions}>
                    {actions.map(ModalBox.renderAction)}
                </div>
            );
        }
    }

    static renderAction(action: Action, index: number) {
        return (
            <button
                key={index}
                className={modalBoxStyles.action}
                onClick={action.handleAction}>{action.title}</button>
        );
    }
}
