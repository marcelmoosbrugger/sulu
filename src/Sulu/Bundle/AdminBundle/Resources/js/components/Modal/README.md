The modal component let's you display some content above everything else.
It renders depending on the passed property and request being closed through a callback.

```
initialState = {open: false};
const actions = [
    {title: 'Destroy world', handleAction: () => {/* destroy world */}},
    {title: 'Save world', handleAction: () => {/* save world */}},
];

<div>
    <button onClick={() => setState({open: true})}>Open modal</button>
    <Modal
        title="Njan Njan Njan"
        onRequestClose={() => setState({open: false})}
        actions={actions}
        confirmText="Apply"
        isOpen={state.open} >
        <div style={{width: '900px', height: '500px', display: 'flex', alignItems: 'center', justifyContent: 'center'}}>
            <img src="http://www.nyan.cat/cats/original.gif" />
        </div>
    </Modal>
</div>
```

If you don't want to handle the open/close state yourself and want to just open the modal
when a specific element is clicked, `ClickModal` is the right choice.
It renders the element which triggers the modal on click and handles the whole opening and closing
of the modal internally.

```
const ClickModal = require('./ClickModal').default;
const actions = [
    {title: 'Save Gotham', handleAction: () => {/* save gotham */}},
];

const button = (<button>Open modal</button>);
<ClickModal
    clickElement={button}
    title="Nana Nana Nana"
    actions={actions}
    confirmText="Ok" >
    <div style={{width: '900px', height: '500px', display: 'flex', alignItems: 'center', justifyContent: 'center'}}>
        <img src="https://media.giphy.com/media/NmhVw98IHkQtq/source.gif" />
    </div>
</ClickModal>
```
