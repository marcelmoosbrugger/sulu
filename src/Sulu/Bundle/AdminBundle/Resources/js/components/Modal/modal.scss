$transitionDuration: 300ms;

.container {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    display: flex;
    justify-content: center;

    & > * {
        opacity: 0;
        transition: opacity $transitionDuration;
    }
}

.box {
    position: relative;
    top: -100%;
    opacity: 0;
    transition: top $transitionDuration, opacity $transitionDuration;
    z-index: 1;
    pointer-events: none;

    & > * {
        pointer-events: auto;
    }
}

.is-down {
    & > * {
        opacity: 1;
    }

    .box {
        top: 0;
    }
}
