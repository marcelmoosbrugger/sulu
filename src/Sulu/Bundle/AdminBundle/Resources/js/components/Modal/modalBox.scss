@import '../../containers/Application/colors.scss';

$primaryBackground: $concrete;
$secondaryBackground: $white;
$borderRadius: 3px;
$boxShadow: 0 10px 18px 0 rgba($mineShaft, .5);

$headerHeight: 60px;
$footerHeight: 90px;

$titleColor: $black;
$titleSize: 22px;

$iconFontSize: 14px;
$iconSize: ($iconFontSize + 10px);

$confirmBackground: $shakespeare;
$confirmColor: $white;
$confirmFontSize: 12px;
$confirmBorderRadius: 3px;
$confirmHeight: 30px;

.box {
    border-radius: 0 0 $borderRadius $borderRadius;
    overflow: hidden;
    box-shadow: $boxShadow;

    header {
        background: $secondaryBackground;
        height: $headerHeight;
        color: $titleColor;
        font-size: $titleSize;
        font-weight: bold;
        text-align: center;
        line-height: $headerHeight;
    }

    article {
        max-height: calc(100vh - $headerHeight - $footerHeight);
        overflow: auto;
        background: $primaryBackground;
    }

    footer {
        background: $secondaryBackground;
        height: $footerHeight;
        padding: 0 60px;
        display: flex;
        justify-content: space-between;
        align-items: center;
    }

    .icon {
        position: absolute;
        top: calc(($headerHeight - $iconSize) / 2);
        right: 30px;
        font-size: $iconFontSize;
        width: calc($iconSize);
        height: calc($iconSize);
        line-height: calc($iconSize);
        text-align: center;
        cursor: pointer;
    }
}

.confirm-button {
    background: $confirmBackground;
    color: $confirmColor;
    font-size: $confirmFontSize;
    border: none;
    border-radius: $confirmBorderRadius;
    line-height: $confirmHeight;
    cursor: pointer;
    text-align: center;
    min-width: 140px;
    padding: 0 40px;
}
