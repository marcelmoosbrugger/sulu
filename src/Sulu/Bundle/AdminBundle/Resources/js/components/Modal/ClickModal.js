// @flow
import React from 'react';
import type {Element} from 'react';
import {action, observable} from 'mobx';
import {observer} from 'mobx-react';
import type {ModalProps} from './types';
import Modal from './Modal';

type Props = ModalProps & {
    className?: string,
    clickElement: Element<*>,
};

@observer
export default class ClickModal extends React.PureComponent<Props> {
    @observable modalOpen = false;

    @action handleElementClick = () => {
        this.modalOpen = true;
    };

    @action handleRequestClose = () => {
        this.modalOpen = false;
    };

    render() {
        const {className, clickElement, ...modalProps} = this.props;
        return (
            <div className={className}>
                {React.cloneElement(clickElement, {onClick: this.handleElementClick})}
                <Modal
                    isOpen={this.modalOpen}
                    onRequestClose={this.handleRequestClose}
                    {...modalProps}>
                    {this.props.children}
                </Modal>
            </div>
        );
    }
}
