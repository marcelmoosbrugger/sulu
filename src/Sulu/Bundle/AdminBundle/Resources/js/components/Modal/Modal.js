// @flow
import React from 'react';
import type {Node} from 'react';
import {observable, action} from 'mobx';
import {observer} from 'mobx-react';
import Portal from 'react-portal';
import classNames from 'classnames';
import Icon from '../Icon';
import {afterElementsRendered} from '../../services/DOM';
import Backdrop from '../Backdrop';
import type {Action} from './types';
import Actions from './Actions';
import modalStyles from './modal.scss';
import modalBoxStyles from './modalBox.scss';

type Props = {
    title: string,
    children: Node,
    actions: Array<Action>,
    confirmText: string,
    onConfirm: () => void,
    isOpen: boolean,
    onRequestClose: () => void,
};

const CLOSE_ICON = 'times';

@observer
export default class Modal extends React.PureComponent<Props> {
    static defaultProps = {
        isOpen: false,
        actions: [],
    };

    @observable isVisible: boolean = false;
    @observable isOpenHasChanged: boolean = false;

    @action componentWillMount() {
        this.isOpenHasChanged = this.props.isOpen;
    }

    componentDidMount() {
        this.toggleModal();
    }

    @action componentWillReceiveProps(newProps: Props) {
        this.isOpenHasChanged = newProps.isOpen !== this.props.isOpen;
    }

    componentDidUpdate() {
        this.toggleModal();
    }

    close() {
        this.props.onRequestClose();
    }

    @action toggleModal() {
        afterElementsRendered(action(() => {
            if (this.isOpenHasChanged) {
                this.isVisible = this.props.isOpen;
            }
        }));
    }

    @action handleTransitionEnd = () => {
        afterElementsRendered(action(() => {
            this.isOpenHasChanged = false;
        }));
    };

    handleIconClick = () => {
        this.close();
    };

    render() {
        const containerClass = classNames({
            [modalStyles.container]: true,
            [modalStyles.isDown]: this.isVisible,
        });
        const {isOpen, title, actions, onConfirm, confirmText, children} = this.props;

        return (
            <Portal isOpened={isOpen || this.isOpenHasChanged}>
                <div
                    className={containerClass}
                    onTransitionEnd={this.handleTransitionEnd}>
                    <div className={modalStyles.box}>
                        <section className={modalBoxStyles.box}>
                            <header>
                                {title}
                                <Icon
                                    name={CLOSE_ICON}
                                    className={modalBoxStyles.icon}
                                    onClick={this.handleIconClick} />
                            </header>
                            <article>{children}</article>
                            <footer>
                                <Actions actions={actions} />
                                <button className={modalBoxStyles.confirmButton} onClick={onConfirm}>
                                    {confirmText}
                                </button>
                            </footer>
                        </section>
                    </div>
                    <Backdrop local={true} onClick={this.props.onRequestClose} />
                </div>
            </Portal>
        );
    }
}
