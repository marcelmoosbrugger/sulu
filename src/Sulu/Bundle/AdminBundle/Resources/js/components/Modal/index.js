// @flow
import ClickModal from './ClickModal';
import Modal from './Modal';

export {ClickModal};
export default Modal;
