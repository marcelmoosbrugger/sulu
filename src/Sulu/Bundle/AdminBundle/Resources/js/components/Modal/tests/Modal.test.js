/* eslint-disable flowtype/require-valid-file-annotation */
import {mount, shallow} from 'enzyme';
import React from 'react';
import pretty from 'pretty';
import Modal from '../Modal';

afterEach(() => document.body.innerHTML = '');

test('The component should render in body when open', () => {
    const body = document.body;
    const onRequestClose = () => {};
    const view = mount(
        <Modal
            title="My modal title"
            onRequestClose={onRequestClose}
            confirmText="Apply"
            isOpen={true} >
            <p>My modal content</p>
        </Modal>
    ).render();
    expect(view).toMatchSnapshot();
    expect(pretty(body.innerHTML)).toMatchSnapshot();
});

test('The component should not render in body when closed', () => {
    const body = document.body;
    const onRequestClose = () => {};
    const view = mount(
        <Modal
            title="My modal title"
            onRequestClose={onRequestClose}
            confirmText="Apply"
            isOpen={false} >
            <p>My modal content</p>
        </Modal>
    ).render();
    expect(view).toMatchSnapshot();
    expect(body.innerHTML).toBe('');
});

test('The component should request to be closed on click on backdrop', () => {
    const requestCloseSpy = jest.fn();
    const view = shallow(
        <Modal
            title="My modal title"
            onRequestClose={requestCloseSpy}
            confirmText="Apply"
            isOpen={true} >
            <p>My modal content</p>
        </Modal>
    );
    const backdrop = view.find('Backdrop');
    expect(backdrop.length).toBe(1);

    expect(requestCloseSpy).not.toBeCalled();
    backdrop.props().onClick();
    expect(requestCloseSpy).toBeCalled();
});
