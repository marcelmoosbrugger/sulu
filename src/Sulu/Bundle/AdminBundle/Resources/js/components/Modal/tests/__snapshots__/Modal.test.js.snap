// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The component should not render in body when closed 1`] = `
<div>
  
</div>
`;

exports[`The component should render in body when open 1`] = `
<div>
  
</div>
`;

exports[`The component should render in body when open 2`] = `
"<div>
  <div data-reactroot=\\"\\" class=\\"container\\">
    <div class=\\"box\\">
      <div class=\\"box\\">
        <div class=\\"header\\">
          <!-- react-text: 5 -->My modal title
          <!-- /react-text --><span class=\\"icon fa fa-times\\" aria-hidden=\\"true\\"></span></div>
        <div class=\\"content\\">
          <p>My modal content</p>
        </div>
        <div class=\\"footer\\">
          <button class=\\"confirmButton\\">Apply</button>
        </div>
      </div>
    </div>
    <div class=\\"backdrop isVisible\\"></div>
  </div>
</div>"
`;
